### DISCUSSION TITLE
Add download zip option for shared link album downloads

### DISCUSSION THREAD

**Discussion by mb0x88** [ente-io/ente#628] [Enhancements]
![Screenshot 2023-11-08 at 21 31 15](https://github.com/ente-io/photos-desktop/assets/106029773/92043943-5c3e-4cff-8025-2d16e361b2a3)

Download album downloads each photo file by file, that's cool but would be great to have an option to download the whole album in a zip folder.

**Comment by vishnukvmd** [comment]
This could be useful - [jszip](https://www.npmjs.com/package/jszip)

**Comment by fahrstuhl** [comment]
Just ran into this. Downloading file by file is a bad user experience when sharing a link to an ente album with other people. If your browser is set to ask everytime where to save a file, at least Firefox proceeds to open a dialog for every single photo. Which is unusable for albums with thousands of photos.

**Comment by Phantom0025** [comment]
This would be a cool feature. How likly is it to get implemented (soon)?

**Comment by Gong-Bao-Chicken** [comment]
I shared a link for an album from ente with friends, they were quite surprised they had hundres of pictures in their download folder thereafter.

+ 1 For implementing download as zip, if either multiple images are selected or if they choose to download all images from dropdown.

**Comment by danlamanna** [comment]
FWIW, this issue is particularly bothersome to me so I decided I would take a stab at it. Unfortunately it ended up being far more complicated than I anticipated so I'm abandoning it for now, but I figured I would dump some of my thoughts here for anyone in the future.

Given the data has to be decrypted on the client side and the size of the zip could be larger than memory, everything has to be done in a streaming fashion. JS in the browser doesn't currently have the ability to write to the user's filesystem (see https://fs.spec.whatwg.org/ for the spec that will eventually enable this), so the only way to currently provide this is to intercept an HTTP request from an `a` tag and generate the zip entirely in a service worker. This repo provides a nice demo of this workflow: https://github.com/Touffy/client-zip/blob/7fcc01b4b72f8bf0c8e9e26b5d29e5c6b22faab0/demo/worker.html. 

Unfortunately this runs into at least a couple of complications: 

1) How do images get decrypted in the service worker - do the bytes get message passed across from the main thread, or does the service worker handle calling out to the decryption/fetch methods?
2) If non-service worker threads are involved, how can a zip download successfully if the user closes the tab once the download has started?

I'm pretty rusty with JavaScript and have already sunk a bit more time into this than I had wanted, but hopefully this helps anyone who might want to go forward with this in the future.

**Reply by danlamanna** [reply]
To update this - I actually think a better middleground now is to just allow the browsers that support the [File System Access API](https://developer.chrome.com/docs/capabilities/web-apis/file-system-access)[^1] to download the zip directly to their filesystem. This can be all be done in a *web* worker by preventing the user from closing the tab during the download. The implementation of this sort of feature is significantly more straightforward than what I previously recommended.

[^1]: "The File System Access API is supported on most Chromium browsers on Windows, macOS, ChromeOS, Linux, and Android. A notable exception is Brave where it is [currently only available behind a flag](https://github.com/brave/brave-browser/issues/18979)." Firefox is on record stating that they will not support this API.

**Reply by codethief** [reply]
@danlamanna I don't know much about service workers but after a bit of research I came across [StreamSaver.js](https://github.com/jimmywarting/StreamSaver.js), specifically [this example](https://github.com/jimmywarting/StreamSaver.js/blob/master/examples/saving-multiple-files.html), which seems to be more or less exactly what we want to do here.

**Comment by 3cHeLoN** [comment]
This feature is essential. In firefox, the popup message that the download is ready steals focus. This makes using the browser impossible as every half second when a new photo is downloaded, you lose focus on the field you are typing.
I think google drive has a similar option. If the zip-file is too large you can create a multi-file zip file with a size of half a GB or something like that.

**Comment by Yolakalemowa** [comment]
Urgently needed feature :(

**Comment by KeebleGH** [comment]
Yeah we really need this, how this isnt a feature is actually insane

**Comment by AswinAsok** [comment]
hi folks,

the download as zip feature is now live on the web client - #627 

<img width="402" height="297" alt="image" src="https://github.com/user-attachments/assets/1ddbbad4-43cd-46a8-81e0-1e688fbd11e0" />

**Reply by mo-st** [reply]
Hi there and thanks for the effort you put into this. I just tried this out and my honest take is that this just reduces the problem slightly, it doesn't actually fix it. Sorry to say, but I'm a bit disappointed. 

For people reading who didn't try this, this is the new behavior on mobile when downloading from a shared link:
Instead of downloading each photo indvidually, it now packs about 4 photos into zip. For example, for relatively small amount of 38 photos, this will still create 10 individual zip files, each downloading in sequence and requiring manual confirmation. 
For smaller amounts of files this is a small improvement, but when sharing hundreds of files this is still basically unsusable as a feature.
Especially annoying is the fact that even after canceling the download the download popups may continue for minutes. 

I suspect the reason for this are the difficulties listed above and maybe the behavior is a bit better than with more memory. Still, this is no solution to the problem in my opinion.

Maybe I misunderstood your usecase, but as I understand the problem, this doesn't solve it.

**Reply by AswinAsok** [reply]
hi @mo-st

sorry to hear that. i actually spent a few weeks trying my best to achieve the ideal case - downloading all photos as a single ZIP - but couldn't arrive at a solution that ensured both cross-browser compatibility and was fail-proof.

so eventually i had to go with the multiple zip method.

also, for the size of photos downloading as multiple zips, we've kept a memory cap to ensure the process doesn't eat up the whole browser memory. currently it's ~100MB for mobile and ~250MB for browsers.

i plan to increment this slowly as we don't know the max limit that works on all devices yet, so went with the safest numbers for now.

regarding popups continuing after canceling - can you give me a bit more clarification on the current behavior? when i tried it just now, the popup closed instantly without any further messages.

**Reply by mo-st** [reply]
hi @AswinAsok 
Thanks for your quick answer! 

Hmm I see. So you choose to rather implement something stable and universal than something fragile that solves the problem only in a few scenarios. That is a valid choice and I'd have done the same, thank you for explaining

Okay, that makes sense, so maybe in the future the improvement will be felt more. And I didn't properly acknowledge this but you did of course improve the situation, I'm sorry if I came off as entitled. I think it is just frustrating to me because I love your product and values but most people don't appreciate me sharing photos with them through ente because of this, and often actually prefer me sending them through some messenger even at a loss of quality because they feel the download is somehow broken.
Still, I do appreciate your work and effort in trying to improve it, even if it's not quite your (and my) ideal solution.

Of course, I'd love to. Using Chrome on an Android 16 Pixel, when tapping the download all icon in the shared album my observed behavior is this:
1. Info toast from ente itself pops up in the bottom left corner of the website, preparing the download
2. As soon as 4 zips are prepared, the browser itself pops up the download dialogue. I tap ok, selecting the default download folder
3. As this happens again, another browser popup opens, asking if I want to allow the site to download multiple files at once. I tap ok
4. At some point I tap cancel in the browser dialogue and the "x" of the ente info window about the zipping progress
5. ente pops up a confirmation dialogue, asking if I'm sure I'd like to stop the download. I confirm via the red button
6. Another browser download dialogue opens, which I cancel too
7. I lock the phone without closing the tab or leaving the browser
8. I unlock the phone and another download popup opens. I cancel

Maybe this is some strange edge case because I immediately locked the phone. Anyway it's not so important I think if you intend to iteratively improve the bundling anyway. At least I now know that this is a technical limitation and that it just doesn't seem possible to get a single download.

I was just thinking of other E2E encrypted web downloads, would it be possible if the photos where combined in a single encrypted zip on my side when I *create* the share? I don't know all the ins-and-outs but in my basic understanding the heart of the issue is that you have individual encrypted files and you can't bundle them to a single file in this state, right? Instead of having each recipient of the share carry the burden, I think it would be nice to have it instead take a moment when creating the share on my device to bundle them all together, even if that means increased waiting time and storage from my plan used.

**Reply by AswinAsok** [reply]
hi @mo-st,

i acknowledge this isn't the best solution, but as you mentioned, it's indeed better than the previous one.
regarding the browser popups you mentioned in points 2, 3 (from browser) and 5 (Ente) - these popups are expected.

as for the download popup in point 8, i didn't quite get what it was. if you could share a screenshot it would have been great.
also, if it's a notification from the app itself and it's unwanted, we can fix it.

for the solution you proposed at the end - generating the zip file on the user's account and then downloading it. i can't think of any practical implementations for this at the moment, and i assume this would require backend changes as well that would take some time to implement properly.

thanks for the effort in sharing the feedbacks.
and i'll keep this in mind for the next iteration.

**Reply by codethief** [reply]
> would it be possible if the photos where combined in a single encrypted zip on my side when I create the share?

Wouldn't the fundamental issue remain the same? The browser of anyone downloading the encrypted zip file would still have to decrypt it on the fly.

**Reply by codethief** [reply]
Back to the original topic: I agree with @mo-st that the current behavior is frustrating. Personally I'd rather have things work well in a single browser (e.g. Chrome with File System Access API) than in all browsers sub-optimally.

But I'm actually surprised the RAM is presenting an obstacle here. @AswinAsok Have you looked into the link I shared above, https://github.com/jimmywarting/StreamSaver.js ? To me it sounds like it is able to work around the whole RAM issue.

**Reply by AswinAsok** [reply]
Hi @codethief,

The first implementation I tried was using the StreamSaver library and it worked well for Chrome. However, for non-Chromium browsers like Safari and Firefox, the solution wasn't working very well - in the majority of cases, the ZIP file downloaded was corrupted when the process ended and since this solution didn't pass all my test cases, I proceeded with the JSZip method.

**Reply by codethief** [reply]
I see, that's a bummer. Did you raise these issues with the StreamSaver authors? EDIT: I just found [this issue](https://github.com/jimmywarting/StreamSaver.js/issues/366#issue-3846996131) from last week which sounds relevant?

In any case, wouldn't it still make sense to use the library in Chrome, though, to have at least one browser where things work flawlessly? (And possibly add a note in Firefox and Safari.)

**Reply by AswinAsok** [reply]
no, i didn't raise any issues with the StreamSaver authors since bandwidth was limited on my end. after spending a couple of weeks trying to get the best solution, i went with the current implementation.

regarding using StreamSaver for Chrome and jszip for the rest - that's a viable solution, but my thought process then was that i didn't want two implementations for the same feature. wanted to keep it unified as maintaining both implementations would be an overhead.

i'll reconsider this when implementing for desktop(electron), which is a chromium instance.
